testing/citest/tests/kube_smoke_test.py

```diff
@@ -235,6 +234,159 @@ def delete_load_balancer(self):
         self.new_post_operation(
             title='delete_load_balancer', data=payload, path='tasks'),
         contract=builder.build())
+  def create_statefulset_pipeline(self):
+    bindings = self.bindings
+
+    payload = self.agent.make_json_payload_from_kwargs(
+        job=[{
+            'type': 'manual',
+            'user': '[anonymous]'
+        }],
+        description='Create Stateful Sets'.format(
+            self.__lb_name,
+            bindings['SPINNAKER_KUBERNETES_ACCOUNT']),
+        application=self.TEST_APP)
+    builder = kube.KubeContractBuilder(self.kube_observer)
+    (builder.new_clause_builder('Run pipeline', retryable_for_secs=15)
+     .get_resources(
+         'svc',
+         extra_args=['--namespace', self.TEST_NAMESPACE],
+         no_resource_ok=True))
+    return st.OperationContract(
+        self.new_post_operation(
+            title='create_state_ful_sets', data=payload, path='pipelines/'+self.TEST_APP+'/statefulset-pipeline',
+            status_class=st.SynchronousHttpOperationStatus),
+        contract=builder.build())
+
+  def save_statefulset_pipeline(self):
+      bindings = self.bindings
+      payload = self.agent.make_json_payload_from_kwargs(
+			    job=[{
+	                'keepWaitingPipelines': 'false',
+                        'application': self.TEST_APP,
+                        'name': 'statefulset-pipeline',
+	                'lastModifiedBy': 'anonymous',
+	                'limitConcurrent': 'true',
+	                'parallel': 'true',
+	                'stages': [{
+		                'clusters': [{
+			                'account': bindings['SPINNAKER_KUBERNETES_ACCOUNT'],
+			                'application': self.TEST_APP,
+			                'cloudProvider': 'kubernetes',
+			                'containers': [{
+				                    'imageDescription': {
+					                'repository': 'nginx'
+				                        },
+				            'imagePullPolicy': 'ALWAYS',
+				            'limits': {
+					            'memory': '200Mi'
+				                    },
+				            'name': 'nginx',
+				            'requests': {
+					            'cpu': '100m',
+					            'memory': '200Mi'
+				                }
+			                }],
+			        'kind': 'StatefulSet',
+			        'namespace': 'default',
+			        'provider': 'kubernetes',
+			        'region': 'default',
+			        'targetSize': 0,
+			        'terminationGracePeriodSeconds': 30
+		                }],
+		                'name': 'Deploy',
+		                'refId': '1',
+
+		            'type': 'deploy'
+	                }],
+	                'triggers': [{'enabled': 'true'}]
+                }],
+                    description='Save Statefulsets:' ,application=self.TEST_APP)
+      builder = kube.KubeContractBuilder(self.kube_observer)
+      builder.new_clause_builder('Run pipeline', retryable_for_secs=15).get_resources('svc',extra_args=['--namespace', self.TEST_NAMESPACE],no_resource_ok=True)
```

Comment:
I'm not sure what this pipeline is doing because I'm not familiar with the operations. It looks like you dont have a contract here. You are getting resources but then not doing anything with it. You say no_resource_ok but not that you expect no resources. So basically it doesnt matter what happens here, so you havent actually tested anything other than the operation completing. But even then you permit a timeout when you call it.

There's a newer way to test for errors without using no_resource_ok. I havent updated the kube_agent for it (to discourage no_resources_ok and perhaps add explicit support for error checking as I did in gcp and aws agents). I think the most readable way is:

from citest.json_contract import ObservationPredicateFactory
ov_factory = ObservationPredicateFactory()

Then, assuming you are expecting an error:
(builder.new_clause_builder(...)
 .get_resouces('svc', extra_arguments=...)
 .EXPECT(ov_factory.error_list_matches(...))   # or ov_factory.error_list_contains([]))
)

Or if you are expecting it either way, you can add another OR clause with what you would expect.

I think this makes the code and intent clearer as well as clear up the reporting.